<a href="https://colab.research.google.com/github/poudyaldiksha/Data-Science-project/blob/main/Lesson_9_b2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Lesson 9: Numpy, Python Dictionaries

###Activity 1: `reshape()`,` ravel()`,` flatten()`

In [ ]:
array_1 = np.arange(1,10)
array_1

array([1, 2, 3, 4, 5, 6, 7, 8, 9])

In [ ]:
# Activity: Reshape()
import numpy as np
array_1 = np.arange(1,10).reshape(3,3)
array_1

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [ ]:
array_1 = np.arange(1,10).reshape(-1,3)
array_1

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [ ]:
arr_3 = np.arange(1,89).reshape(-1)
arr_3.ndim

1

In [ ]:
# Activity: Ravel()
arr_ravel = array_1.ravel()
arr_ravel[2] = 8989
print(arr_ravel,array_1)

[   1    2 8989    4    5    6    7    8    9] [   1    2 8989    4    5    6    7    8    9]


In [ ]:
#Activity: Flatten()
arr_flatten = array_1.flatten()
arr_flatten[2] = 90000
print(arr_flatten,array_1)

[    1     2 90000     4     5     6     7     8     9] [   1    2 8989    4    5    6    7    8    9]


The `ravel()` and `flatten() `functions in NumPy both return a flattened array, but they have some key differences in terms of how they handle memory:

`ravel()`

- Returns a flattened array.
- Returns a view whenever possible: This means that `ravel()` tries to avoid creating a new array if it can. If the array is already contiguous in memory, it will return a view of the original array. If it can't return a view, it will return a copy.
- Changes to the raveled array may affect the original array if it returns a view.

`flatten()`

- Returns a flattened array.
- Always returns a copy: This means that `flatten()` creates a new array and the changes to the flattened array will not affect the original array.
- Does not affect the original array: Since it returns a copy, any modifications to the flattened array will not impact the original array.

In [ ]:
# Create a 2x2 array
arr = np.array([[1, 2],
                [3, 4]])

# Use ravel() to flatten the array
raveled_arr = arr.ravel()
print("Raveled array:", raveled_arr)

# Modify the raveled array
raveled_arr[0] = 10
print("Modified raveled array:", raveled_arr)
print("Original array after ravel modification:", arr)



Raveled array: [1 2 3 4]
Modified raveled array: [10  2  3  4]
Original array after ravel modification: [[10  2]
 [ 3  4]]


In [ ]:
# Create a 2x2 array
arr = np.array([[1, 2],
                [3, 4]])

# Use flatten() to flatten the array
flattened_arr = arr.flatten()
print("Flattened array:", flattened_arr)

# Modify the flattened array
flattened_arr[0] = 20
print("Modified flattened array:", flattened_arr)
print("Original array after flatten modification:", arr)

Flattened array: [1 2 3 4]
Modified flattened array: [20  2  3  4]
Original array after flatten modification: [[1 2]
 [3 4]]


In [ ]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
arr

array([[1, 2, 3],
       [4, 5, 6]])

In [ ]:
# Create a 2x3 array
arr = np.array([[1, 2, 3], [4, 5, 6]])

# C order (default)
c_order = np.ravel(arr, order='C')
print("C order:", c_order)

# F order
f_order = np.ravel(arr, order='F')
print("F order:", f_order)

C order: [1 2 3 4 5 6]
F order: [1 4 2 5 3 6]


C Order: [1, 2, 3, 4, 5, 6]
- Rows are stored sequentially: first row [1, 2, 3] followed by the second row [4, 5, 6].

F Order: [1, 4, 2, 5, 3, 6]
- Columns are stored sequentially: first column [1, 4], second column [2, 5], and third column [3, 6].

**Practical Use Cases:**

**C Order:**

- Typically used when performing operations that process rows of an array.
- Common in many numerical and scientific computations

**F Order:**

- Useful in linear algebra and matrix operations where column-wise processing is needed.



### Activity 2: Addition of arrays

In [ ]:
# Activity
arr1 = np.arange(1,10).reshape(3,3)
arr1

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [ ]:
arr2 = np.arange(11,20).reshape(3,3)
arr2

array([[11, 12, 13],
       [14, 15, 16],
       [17, 18, 19]])

In [ ]:
arr1 + arr2

array([[12, 14, 16],
       [18, 20, 22],
       [24, 26, 28]])

In [ ]:
np.add(arr1,arr2)

array([[12, 14, 16],
       [18, 20, 22],
       [24, 26, 28]])

In [ ]:
np.subtract(arr1,arr2)

array([[-10, -10, -10],
       [-10, -10, -10],
       [-10, -10, -10]])

In [ ]:
np.multiply(arr1,arr2) # elementwise

array([[ 11,  24,  39],
       [ 56,  75,  96],
       [119, 144, 171]])

In [ ]:
np.divide(arr1,arr2)

array([[0.09090909, 0.16666667, 0.23076923],
       [0.28571429, 0.33333333, 0.375     ],
       [0.41176471, 0.44444444, 0.47368421]])

In [ ]:
# Define two arrays
array1 = np.array([[1, 2, 3], [4, 5, 6]])
array2 = np.array([[7, 8, 9], [10, 11, 12]])

# Add the arrays
result_add = array1 + array2
print("Addition:\n", result_add)

Addition:
 [[ 8 10 12]
 [14 16 18]]


### Activity 3: Subtraction of Arrays

In [ ]:

# Define two arrays
array1 = np.array([[1, 2, 3], [4, 5, 6]])
array2 = np.array([[7, 8, 9], [10, 11, 12]])

# Subtract the arrays
result_subtract = array1 - array2
print("Subtraction:\n", result_subtract)

Subtraction:
 [[-6 -6 -6]
 [-6 -6 -6]]


### Activity 4: Multiplication of Arrays

In [ ]:
import numpy as np
# Define two arrays
array1 = np.array([[1, 2, 3], [4, 5, 6]])
array2 = np.array([[7, 8, 9], [10, 11, 12]])

# Multiply the arrays element-wise
result_multiply = array1 * array2
print("Element-wise Multiplication:\n", result_multiply)

Element-wise Multiplication:
 [[ 7 16 27]
 [40 55 72]]


### Activity 5: Division of Arrays

In [ ]:

# Define two arrays
array1 = np.array([[1, 2, 3], [4, 5, 6]])
array2 = np.array([[7, 8, 9], [10, 11, 12]])

# Divide the arrays element-wise
result_divide = array1 / array2
print("Element-wise Division:\n", result_divide)

Element-wise Division:
 [[0.14285714 0.25       0.33333333]
 [0.4        0.45454545 0.5       ]]


### Activity 6: Array Multiplication (using @ operator)

- The @ operator in Python for numpy arrays is used for matrix multiplication, which is also known as the dot product when applied to 2D arrays (matrices).
- While the dot product of two vectors results in a scalar, the matrix multiplication of two matrices results in another matrix.

Array Multiplication Using @ Operator
In the context of numpy, when you use the @ operator with two 2D arrays (matrices), it performs matrix multiplication.

In [ ]:
# Define two vectors
vector_a = np.array([1, 2, 3])
vector_b = np.array([4, 5, 6])

# Dot product using @ operator
dot_product = vector_a @ vector_b

print("Vector A:", vector_a)
print("Vector B:", vector_b)
print("Dot Product:", dot_product)

Vector A: [1 2 3]
Vector B: [4 5 6]
Dot Product: 32


In [ ]:
# Define two arrays
array1 = np.array([[1, 2], [3, 4], [5, 6]])
array2 = np.array([[7, 8], [9, 10]])

# Matrix multiplication using @ operator
result_matrix_multiplication = array1 @ array2
print("Matrix Multiplication (using @ operator):\n", result_matrix_multiplication)

Matrix Multiplication (using @ operator):
 [[ 25  28]
 [ 57  64]
 [ 89 100]]


In [ ]:
array1 = np.array([[1, 2, 3],
                   [4, 5, 6],
                   [7,8,9]])
array2 = np.array( [[11, 12, 13],
                    [14,15,16],
                    [17,18,19]])
array1 @ array2

array([[ 90,  96, 102],
       [216, 231, 246],
       [342, 366, 390]])

Step-by-Step Calculation

𝐶
11 = (1 * 11) + (2 * 14) + (3 * 17) = 11 + 28 + 51 = 90

Element

𝐶
12 = (1 * 12) + (2 * 15) + (3 * 18) = 12 + 30 + 54 = 96

Element

𝐶
13 = (1 * 13) + (2 * 16) + (3 * 19)= 13 + 32 + 57 = 102

Element

𝐶
21 =(4 * 11) + (5 * 14) + (6 * 17)= 44 + 70 + 102 = 216

Element

𝐶
22 =(4 * 12)+(5 * 15)+(6 * 18)= 48 +75 + 108 =231


Element

𝐶
23 =(4 * 13)+(5 * 16)+(6 * 19)= 52 + 80 + 114 = 246

Element

𝐶
31 =(7 * 11)+(8 * 14)+(9 * 17)= 77 +112 +153 = 342

Element

𝐶
32 =(7 * 12)+(8 * 15)+(9 * 18)= 84 + 120 +162 = 366

Element

𝐶
33 =(7 * 13)+(8 * 16)+(9 * 19)= 91 + 128 + 171 = 390



### Activity 7: `argmax()` and `argmin()`

- The `argmax()`function in NumPy is used to find the index of the maximum value in an array. It can be used with 1-dimensional arrays to return a single index or with multi-dimensional arrays to return indices along a specified axis.

- The `argmin()`function in NumPy is used to find the index of the minimum value in an array.

**1D Array Example**
For a 1D array, argmax() returns the index of the maximum value.

In [ ]:
# For a 1D array, argmax() returns the index of the maximum value.
arr3 =  np.array([1,3,7,2,5])
np.max(arr3)

7

In [ ]:
np.argmax(arr3)

2

In [ ]:
np.argmin(arr3)

0

**2D Array Example**

For a 2D array, argmax() can be used to find the index of the maximum value along a specified axis.

In [ ]:
matrix = np.matrix(" 4 2; 6 -3")
type(matrix)

numpy.matrix

In [ ]:
np.array(" 4 2; 6 -3")

array(' 4 2; 6 -3', dtype='<U10')

In [ ]:
arr = np.array([[1, 2, 3],
                [4, 5, 6],
                 [7, 8, 9]])

# Find the index of the maximum value in the entire array
index_of_max_flat = np.argmax(arr)

# Find the index of the maximum value along each column
index_of_max_axis0 = np.argmax(arr, axis=0)

# Find the index of the maximum value along each row
index_of_max_axis1 = np.argmax(arr, axis=1)

print("Array:")
print(arr)
print("\nIndex of maximum value in the flattened array:", index_of_max_flat)
print("Index of maximum values along axis 0 (columns):", index_of_max_axis0)
print("Index of maximum values along axis 1 (rows):", index_of_max_axis1)

Array:
[[1 2 3]
 [4 5 6]
 [7 8 9]]

Index of maximum value in the flattened array: 8
Index of maximum values along axis 0 (columns): [2 2 2]
Index of maximum values along axis 1 (rows): [2 2 2]


In [ ]:
index_of_max_flat = np.argmax(arr)
index_of_max_flat

In [ ]:
arr.shape

(3, 3)

**Multi-Dimensional Array Example**

For higher-dimensional arrays, argmax() can still be applied along specified axes.

In [ ]:
arr = np.array([[[1, 2, 3], [4, 5, 6]],
               [[7, 8, 9], [10, 11, 12]]])

# Find the index of the maximum value in the entire array
index_of_max_flat = np.argmax(arr)

# Find the index of the maximum value along the first axis
index_of_max_axis0 = np.argmax(arr, axis=0)

# Find the index of the maximum value along the second axis
index_of_max_axis1 = np.argmax(arr, axis=1)

print("Array:")
print(arr)
print("\nIndex of maximum value in the flattened array:", index_of_max_flat)
print("Index of maximum values along axis 0:", index_of_max_axis0)
print("Index of maximum values along axis 1:", index_of_max_axis1)

In [ ]:
arr = np.array([[[1, 2, 3], [4, 5, 6]],
               [[7, 8, 9], [10, 11, 12]]])

# Find the index of the maximum value in the entire array
index_of_max_flat = np.argmax(arr)
index_of_max_flat

In [ ]:
arr = np.array([[[1, 2, 3],
                 [4, 5, 6]],

               [[7, 8, 9],
               [10, 11, 12]]])
index_of_max_axis0 = np.argmax(arr, axis=0)
index_of_max_axis0

In [ ]:
arr_7 = np.array([[[19, 11, 31],
                 [42, 5, 64]],

               [[7, 80, 9],
               [10, 111, 12]]])
index_of_max_axis0 = np.argmax(arr_7, axis=0)
index_of_max_axis0

In [ ]:
arr_7 = np.array([[[19, 11, 31],
                 [42, 5, 64]],

               [[7, 80, 9],
               [10, 111, 12]]])
index_of_max_axis0 = np.argmax(arr_7, axis=1)
index_of_max_axis0

**Summary**

- np.argmax(array) returns the index of the maximum value in a flattened array.
- np.argmax(array, axis=0) returns the indices of the maximum values along the specified axis 0 (columns).
- np.argmax(array, axis=1) returns the indices of the maximum values along the specified axis 1 (rows).

### Activity 8: Concatenate Two array

In [ ]:
# Concatenate using concatenate() function
np.concatenate((arr1,arr2))

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [11, 12, 13],
       [14, 15, 16],
       [17, 18, 19]])

**Row wise concatenate**

In [ ]:
# Activity
np.concatenate((arr1,arr2), axis= 1)

array([[ 1,  2,  3, 11, 12, 13],
       [ 4,  5,  6, 14, 15, 16],
       [ 7,  8,  9, 17, 18, 19]])

**Column wise concatenate**

In [ ]:
#Activity
np.concatenate((arr1,arr2), axis= 0)

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [11, 12, 13],
       [14, 15, 16],
       [17, 18, 19]])

**Concatenate Two array by using Stack function**
`vstack()` and `hstack()` are functions in NumPy used for stacking arrays either vertically or horizontally, respectively.

**`np.vstack()`**

- Function: np.vstack() stacks arrays vertically (row-wise).
- Usage: It takes a sequence of arrays and stacks them vertically to make a single array. The arrays must have the same number of columns (i.e., the same shape along all but the first axis).

In [ ]:
# np.vstack() it will concatenate two array vertically
np.vstack((arr1,arr2))


array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [11, 12, 13],
       [14, 15, 16],
       [17, 18, 19]])

In [ ]:
#Activity

arr1 = np.array([[1, 2, 3],
                 [4, 5, 6]])

arr2 = np.array([[7, 8, 9],
                 [10, 11, 12]])

result = np.vstack((arr1, arr2))

print(result)

[[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]


**`np.hstack()`**

- Function: np.hstack() stacks arrays horizontally (column-wise).
- Usage: It takes a sequence of arrays and stacks them horizontally to make a single array. The arrays must have the same number of rows (i.e., the same shape along all but the second axis).

In [ ]:
# np.hstack() it will concatenate two array horizontally
np.hstack((arr1,arr2))


array([[ 1,  2,  3,  7,  8,  9],
       [ 4,  5,  6, 10, 11, 12]])

In [ ]:
#Activity
arr1 = np.array([[1, 2],
                 [3, 4]])

arr2 = np.array([[5, 6],
                 [7, 8]])

result = np.hstack((arr1, arr2))

print(result)

[[1 2 5 6]
 [3 4 7 8]]


#  Python Dictionaries



**What are Data Structures?**

In Python, data structures are used to store and organize data. They help in managing large amounts of data efficiently. Here are the main types of data structures in Python:

**Built-in Data Structures:**
- List: Ordered, mutable collection of items. Example: [1, 2, 3]
- Tuple: Ordered, immutable collection of items. Example: (1, 2, 3)
- Dictionary: Unordered collection of key-value pairs. Example: {'a': 1, 'b': 2}
- Set: Unordered collection of unique items. Example: {1, 2, 3}

**NOTE**

NumPy arrays are not considered built-in data structures in Python because they are not part of the Python standard library. Instead, they are provided by an external library called NumPy, which needs to be installed separately

### Activity 1: Python Dictionary

Similar to a Python list, a Python dictionary is also a data structure, i.e., it stores multiple values in it. **A python dictionary is a collection of key-value pairs** where each key-value pair collectively constitutes an item stored in it.

So far, you learnt about Python lists in detail. In this class , you will learn about a Python dictionary and its usefulness in terms of data storage.

An example of a key-value pair would be a specification of a laptop, as shown in the table below.

|Key         |Value                  |
|------------|-----------------------|
|Ram         |16 GB                  |
|Storage     |512 GB SSD             |
|Graphics Card|NVIDIA GTX 1650       |
|Display     |15.6 inches            |
|Battery     |4000 mAh               |
|Weight      |1.8 kg                 |
|Price (INR) |96000                  |
|Fast Charge |True                   |
|Processor   |Intel Core i7          |

Using the example shown in the table here, let's create a dictionary as a collection of key-value pairs.

A dictionary is created using curly brackets (`{}`). A key and its corresponding value are mapped using the colon (`:`) sign and each key-value pair is separated by a comma.

In [ ]:
a ={-1:"negative",1:"positive",0:"zero"}
a

{-1: 'negative', 1: 'positive', 0: 'zero'}

In [ ]:
type(a)

dict

In [ ]:
# Activity: Create a sample dictionary
diet = { "breakfast" : " dosa", "lunch" : " roti", "dinner " : "rice"}
diet


{'breakfast': ' dosa', 'lunch': ' roti', 'dinner ': 'rice'}

In [ ]:
type(diet)

dict

In [ ]:
student = { "name" : "john",
             "age": 25,
             "city": "newyork"}
student

{'name': 'john', 'age': 25, 'city': 'newyork'}

In [ ]:
#Create a dictionary as a collection of key-value pairs containing the items shown in the table above.

laptop_specs = {
    "Ram": "16 GB",# First key-value pair
    "Storage": "512 GB SSD",# Second key-value pair
    "Graphics Card": "NVIDIA GTX 1650",# Third key-value pair
    "Display": "15.6 inches",# Fourth key-value pair
    "Battery": "4000 mAh", # Fifth key-value pair
    "Weight": "1.8 kg",# Sixth key-value pair
    "Price (INR)": 96000,# Seventh key-value pair
    "Fast Charge": True,# Eighth key-value pair


}

# Printing he dictionary
laptop_specs

{'Ram': '16 GB',
 'Storage': '512 GB SSD',
 'Graphics Card': 'NVIDIA GTX 1650',
 'Display': '15.6 inches',
 'Battery': '4000 mAh',
 'Weight': '1.8 kg',
 'Price (INR)': 96000,
 'Fast Charge': True}

In [ ]:

laptop_specs = {
    "Ram": "16 GB",
    "Storage": "512 GB SSD",
    "Graphics Card": "NVIDIA GTX 1650",
    "Display": "15.6 inches",
    "Battery": "4000 mAh",
    "Weight": "1.8 kg",
    "Price (INR)": 96000,
    "Fast Charge": True,


}

In the above code, we have created a Python dictionary and stored it in `laptop_specs` variable. The dictionary contains 8 key-value pairs separated by a comma. The keys are of `string` data-type and the values are a mix of `string`, `int` and `bool` data types.

**Note:**
-  Even though you store an item (a key-value pair) in a specific order, a Python dictionary disregards the order. Hence, unlike a Python list, a Python dictionary is a data structure containing an **unordered** collection of items.
-  lists maintain order based on their indices, dictionaries do not guarantee any specific order for their elements.
-  It means that you can't retreive an item from a dictionary using indices.




Behavior in Google Colab (or similar environments):

- In environments like Google Colab, you might notice that dictionaries sometimes appear to maintain the order in which items were added.
- This is due to how the environment or specific Python versions implement dictionaries. It can give the impression of order, but this behavior is not guaranteed in all Python environments or versions.

**Important Note:**

- When learning and using dictionaries, it’s important to understand and expect this unordered behavior, even if specific environments may show them differently.


### Activity 2: Empty Python Dictionary

You can also create an empty dictionary$—$a dictionary which does not have any key-value pair$—$by simply writing curly brackets `{}` only.

In [ ]:
# Create an empty dictionary.
emp_dict = {}
emp_dict

{}

In [ ]:
a =(1,2,3)
a

(1, 2, 3)

### Activity 3: Important Points About Python Dictionary

Here are some of the important points about a Python dictionary:

1. A key in a Python dictionary can be a `str` (string), `int` (integer), `float` (a floating-point number) and a `bool` (boolean) data-type. Additionally, it can also be a tuple. In general, any immutable Python object can be a key in a Python dictionary. The term **immutable** means **unchangeable**.

2. Any Python object (whether mutable or immutable) or a data-type can be a value in a Python dictionary.

Let's go through a few examples for keys and values in a Python dictionary.

**Booleans as keys**

In [ ]:
# Create a Python dictionary whose keys are 'True' and 'False'. You can assign them any values.
bool_as_keys = { True : "yes ", False :"no"}
bool_as_keys

{True: 'yes ', False: 'no'}

**Float-point numbers as keys**

In [ ]:
#Create a Python dictionary whose keys are some floating-point numbers. You can assign them any values.
float_as_key = {3.14:"pi",2.71:"e"}
float_as_key

{3.14: 'pi', 2.71: 'e'}

**Integers as keys**

In [ ]:
#Create a Python dictionary whose keys are integer values. You can assign them any values.
i = { 1:"one",2:"two"}
i

{1: 'one', 2: 'two'}

**Tuples as keys**

In [ ]:
# Create a Python dictionary whose keys are some tuples. You can assign them any values.
t = {(2,4,6):"even",(3,5,7):"odd"}
t

{(2, 4, 6): 'even', (3, 5, 7): 'odd'}

In [ ]:
#t = {[2,4,6]:"even",[3,5,7]:"odd"} # List cannot be a key
#t

**Lists as values**

In [ ]:
# Create a Python dictionary whose values are some lists.
l = { "prime": [2,3,5,7,11],"odd":[1,3,5,7,9]}
l

{'prime': [2, 3, 5, 7, 11], 'odd': [1, 3, 5, 7, 9]}

**Dictionaries as values**

In [ ]:
diet

{'breakfast': ' dosa', 'lunch': ' roti', 'dinner ': 'rice'}

In [ ]:
# Create a Python dictionary whose values are some other dictionaries.
a = {"rina": diet, "john": {"a": 34,"b":45}}
a

{'rina': {'breakfast': ' dosa', 'lunch': ' roti', 'dinner ': 'rice'},
 'john': {'a': 34, 'b': 45}}

**Tuples as values**

You will learn about tuples in detail in the upcoming classes.

In [ ]:
# Create a Python dictionary whose values are tuples.
l = { "prime": (2,3,5,7,11),"odd":(1,3,5,7,9)}
l

{'prime': (2, 3, 5, 7, 11), 'odd': (1, 3, 5, 7, 9)}

**Note:**

- A **key** in a dictionary must be unique. In other words, two items in a dictionary cannot have the same **key**.
- If a dictionary contains two items having the same key, then the item at the **higher index** will be accepted by the dictionary and the item at the **lower index** will be discarded.


In [ ]:
# Create a dictionary having two identical keys but different values.
dict_with_identical_keys = {'Super': 'Man', 'Hot' : 'Chocolate', 'Super' : 'Woman'}
dict_with_identical_keys

{'Super': 'Woman', 'Hot': 'Chocolate'}

In [ ]:
#Activity: values are identical
dict_with_identical_keys = {'Apple': 'Fruit', 'Apple': 'Tech Company', 'Banana': 'Fruit'}
dict_with_identical_keys

{'Apple': 'Tech Company', 'Banana': 'Fruit'}

As you can see, the latest value for a key overrides its previous value.

**Note:** Two different keys in a dictionary can have the same value.

In [ ]:
# Create a dictionary having two identical values but different keys.
dict_with_identical_values = {'Super' : 'Man', 'Hot' : 'Chocolate', 'Bat' : 'Man'}
dict_with_identical_values

{'Super': 'Man', 'Hot': 'Chocolate', 'Bat': 'Man'}

In [ ]:
#Activity
dict_with_identical_values = {'Apple': 'Fruit', 'Orange': 'Fruit', 'Banana': 'Fruit'}
dict_with_identical_values


{'Apple': 'Fruit', 'Orange': 'Fruit', 'Banana': 'Fruit'}

### Activity 4: Dictionary Length
The length of a dictionary is the number of key-value pairs contained in the dictionary.

To calculate the length of a dictionary (or number of key-value pairs contained in a dictionary), you can use the `len()`  function.

In [ ]:
# Calculate the length of the 'laptop_specs' dictionary.
len(laptop_specs)

---

###Activity 5: The `keys()` Function
To get a list of all the keys in a dictionary, you can use the `keys()` function using the `dictionary_name.keys()` method.



In [ ]:
#Get the list of all the keys contained in 'laptop_specs' dictionary.
s= laptop_specs.keys()
type(s)



**Note** that the list returned by the `keys()` function is not the same as a Python list. It behaves differently as compared to a conventional Python list.

You can also use a `for` loop to print all the keys of a dictionary one after the other.


In [ ]:
# Get the list of all the keys contained in 'laptop_specs' dictionary using 'for' loop.
for i in laptop_specs.keys():
    print(i)

Actually, even if you don't use the `keys()` function, you can retrieve all the keys from a Python dictionary using a `for` loop. Let's retrieve all the keys from the `laptop_specs` dictionary without using the `keys()` function.

In [ ]:
# Retrieve all the keys present in 'laptop_specs' dictionary without using the 'keys()' function but using 'for' loop.
for i in laptop_specs:
    print(i)

---

### Activity 6: Getting A Value For A Key

Now let's learn how to retrieve a value for a key from a Python dictionary. There are two methods to do this. They are:

1. The square brackets method

2. The `get()` function


**Method I: The Square Brackets**

For a key, in a dictionary, you can retrieve its corresponding value by using the `dictionary_name[key]` method where `dictionary_name` is the variable storing a Python dictionary and `key` is the key stored in the dictionary.


In [ ]:
laptop_specs["Ram"]

In [ ]:
# Print the corresponding values to each key contained in the 'my_dict' dictionary.

# Value for the first key.
print(laptop_specs['Ram'])

# Value for the second key.
print(laptop_specs['Storage'])

# Value for the third key.
print(laptop_specs['Graphics Card'])

print(laptop_specs['Display'])

# Value for the fifth key.
print(laptop_specs['Battery'])
# Value for the sixth key.
print(laptop_specs['Weight'])

# Value for the seventh key.
print(laptop_specs['Price (INR)'])

# Value for the eighth key.
print(laptop_specs['Fast Charge'])



Instead of retrieving all the values one-by-one, you can get them using a `for` loop along with keys enclosed in square brackets.

In [ ]:
# Print all the values contained in the 'laptop_specs' dictionary using a 'for' loop, 'keys()' function & square brackets.
for i in laptop_specs.keys():
    print(laptop_specs[i])

In [ ]:
# Print all the values stored in the 'laptop_specs' dictionary using 'for' loop & the square brackets.
for i in laptop_specs:
    print(laptop_specs[i])


**Note:** In a way, the keys act like indices for a Python dictionary. Unlike a Python list, a Python dictionary does not have indices of their own because a dictionary is an **unordered** collection of items. An only ordered collection of items have indices. Also, since a Python dictionary does not have indices, it is best to use `for` loop over `while` loop to iterate through each item that is stored in a dictionary.

**Method II: The `get()` Function**

You can also use the `get()` function to get a value of a key by using the `dictionary_name.get(key)` method.


In [ ]:
# Print the corresponding values to each key contained in the 'my_dict' dictionary using the 'get()' function.

# Value for the first key.
print(laptop_specs.get('Ram'))

# Value for the second key.
print(laptop_specs.get('Storage'))

# Value for the third key.
print(laptop_specs.get('Graphics Card'))



Similarly, you can also use the `get()` function with `for` loop to get the value for a key.

In [ ]:
# Print all the values contained in the 'laptop_specs' dictionary using 'for' loop and the 'get()' function.
for i in laptop_specs.keys():
    print(laptop_specs.get(i))

In the code above, `for` loop iterates through each key in the list generated by the `keys()` function. Then, the `get()` function returns the value for a key. Finally, the `print()` function prints the corresponding value for each key.

---

### Activity 7: The `values()` Function

Let's learn how to get the collection of all the values contained in a Python dictionary. You can do this by using the `values()` function.

**Syntax:** `dictionary_name.values()`


In [ ]:
#  Retrieve a collection of all the values stored in the 'laptop_specs' dictionary using the 'values()' function.
laptop_specs.values()

As you can see, the `values()` function has returned a list of all the values contained in the `my_dict` dictionary.

**Note** that the list returned by the `values()` function is not the same as a Python list. This list behaves differently as compared to a conventional Python list.


---

### Activity 8: The `items()` Function

You can also retrieve all the key-value pairs from a dictionary as a collection of tuples by using the `items()` function.

**Syntax:** `dictionary_name.items()` method.


In [ ]:
# Retrieve all the items stored in the 'laptop_specs' dictionary as a collection of tuples.
laptop_specs.items()

As you can see, the `items()` function has returned a list of key-value pairs contained in the dictionary. Each key-value pair is a tuple because a pair is written inside the common brackets, i.e., `()`.

**Note** that the list returned by the `items()` function is not the same as a Python list. This list behaves differently as compared to a conventional Python list.

You can also use a `for` loop to print each item contained in the list returned by the `items()` function.


In [ ]:
# Loop through each item stored in the list returned by the 'items()' function.
for item in laptop_specs.items():
  print(item)

As you can see, using `for` loop, we have retrieved every tuple containing a key-value pair, from the `laptop_specs` dictionary.

You can also simultaneously iterate through each key-value contained in the list returned by the `items()` function.

In [ ]:
for i,j in laptop_specs.items():
    print(i,j)

In [ ]:
# Simultaneously iterate through each key-value pair contained in the list returned by the 'items()' function.
for key, value in laptop_specs.items():
  print(key, value)

As you can see, in the `for` loop, the `key` variable iterates through each key in each item and the `value` variable iterates through each value in each item contained in the list,


Then, the `print()` function prints the values stored in the `key` and `value` variables.

If you apply the same method of retrieving key-value pairs simultaneously directly from the `laptop_specs` dictionary, then you will get `ValueError`.

In [ ]:
#Simultaneously retrieve all the key-value pairs directly from the 'laptop_specs' dictionary.
for key in laptop_specs:
  print(key)

Hence, to simultaneously iterate through multiple items (or key-value pairs), it is best to have each item (or key-value pair) as a collection of tuples. You will learn about tuples in detail in the upcoming classes.


### Activity 9: Adding New Item

Let's also add the Processor of the laptop to the `laptop_specs` dictionary. So the new item is `"Processor"` and the corresponding value is `'Intel Core i7'`.

To add a new item to a dictionary, you simply have to use the square brackets as shown in the syntax below.

**Syntax:** `dictionary_name[key] = value`

where `dictionary_name` is some variable storing a dictionary.

In [ ]:
# Add 'processor' and 'Intel Core i7' as keys and values respectively to the 'laptop_specs' dictionary.
laptop_specs["processor"] = "Intel Core i7"
laptop_specs

---